File: dart/lib/src/utils/reverse_lookup_batch.dart
import '../rpc/rpc_client.dart';
import '../states/registry.dart';
import 'deserialize_reverse.dart';
import 'get_reverse_address_from_domain_address.dart';

/// Parameters for performing batch reverse lookup
class ReverseLookupBatchParams {
  const ReverseLookupBatchParams({
    required this.rpc,
    required this.domainAddresses,
  });

  /// The RPC client to interact with the blockchain
  final RpcClient rpc;

  /// The domain addresses to perform the reverse lookup on
  final List<String> domainAddresses;
}

/// Perform a batch reverse lookup for given domain addresses.
///
/// This function mirrors js-kit/src/utils/reverseLookupBatch.ts
///
/// [params] - Parameters containing RPC client and domain addresses
///
/// Returns a promise that resolves to a list of human-readable domain names
/// associated with the given addresses, or null if the data is not available.
Future<List<String?>> reverseLookupBatch(
    ReverseLookupBatchParams params) async {
  final reverseLookupAddresses = await Future.wait(
    params.domainAddresses.map(getReverseAddressFromDomainAddress),
  );

  final states =
      await RegistryState.retrieveBatch(params.rpc, reverseLookupAddresses);

  return states
      .map((state) =>
          state?.data != null ? deserializeReverse(state!.data) : null)
      .toList();
}

File: js/src/utils/reverseLookupBatch.ts
import { Connection, PublicKey } from "@solana/web3.js";
import { NameRegistryState } from "../state";
import { REVERSE_LOOKUP_CLASS } from "../constants";

import { getHashedNameSync } from "./getHashedNameSync";
import { getNameAccountKeySync } from "./getNameAccountKeySync";
import { deserializeReverse } from "./deserializeReverse";

/**
 * Performs reverse lookups for domain accounts.
 *
 * @param connection Solana RPC connection
 * @param nameAccounts Domain account public keys to reverse look up
 * @returns Human-readable domain names when reverse accounts exist.
 */
export async function reverseLookupBatch(
  connection: Connection,
  nameAccounts: PublicKey[],
): Promise<(string | undefined)[]> {
  let reverseLookupAccounts: PublicKey[] = [];
  for (let nameAccount of nameAccounts) {
    const hashedReverseLookup = getHashedNameSync(nameAccount.toBase58());
    const reverseLookupAccount = getNameAccountKeySync(
      hashedReverseLookup,
      REVERSE_LOOKUP_CLASS,
    );
    reverseLookupAccounts.push(reverseLookupAccount);
  }

  let names = await NameRegistryState.retrieveBatch(
    connection,
    reverseLookupAccounts,
  );

  return names.map((name) => {
    if (name === undefined || name.data === undefined) {
      return undefined;
    }
    return deserializeReverse(name.data);
  });
}
